File: AtCoder/ABC/ABC101-150/abc120/abc120d.py
class UnionFind():

  def __init__(self, n):
    self.par = [i for i in range(n)]
    self.rank = [0] * n
    self.sizes = [1] * n

  def find(self, x):
    if self.par[x] != x:
      self.par[x] = self.find(self.par[x])
    return self.par[x]

  def union(self, x, y):
    x, y = self.find(x), self.find(y)
    if x == y:
      return False
    elif self.rank[x] < self.rank[y]:
      self.par[x] = y
      self.sizes[y] += self.sizes[x]
    elif self.rank[x] > self.rank[y]:
      self.par[y] = x
      self.sizes[x] += self.sizes[y]
    else:
      self.par[y] = x
      self.rank[x] += 1
      self.sizes[x] += self.sizes[y]
    return True

  def same(self, x, y):
    return self.find(x) == self.find(y)

  def size(self, x):
    return self.sizes[self.find(x)]

n, m = list(map(int, input().split()))
edge = [tuple(map(int, input().split())) for i in range(m)]
ans = []
uf = UnionFind(n+1)
for x, y in edge[::-1]:
  if uf.same(x, y):
    ans.append(0)
    continue
  else:
    ans.append(uf.size(x)*uf.size(y))
    uf.union(x, y)
cur = 0
for e in ans[::-1]:
  cur += e
  print(cur)

File: AtCoder/ABC/ABC101-150/abc120/abc120d.cpp
#include <bits/stdc++.h>
using namespace std;
#define REP(i,b,e) for(int i=(b);i<(e);++i)
#define RREP(i,b,e) for(int i=(b)-1;i>=e;--i)
#define rep(i,e) for(int i=0;i<(e);++i)

inline void print(void) { cout<<'\n'; }
template<class T> inline void print(const T &x) { cout<<x<<'\n'; }
template<class T, class... U> inline void print(const T &x, const U&... y) { cout<<x<<" "; print(y...); }

struct UnionFind{
private:
  vector<int> par, rank, sizes;
public:
  UnionFind(int n) : par(n, -1), rank(n, 0), sizes(n, 1){}
  int find(int x) { return (par[x] < 0) ? x : (par[x] = find(par[x])); }
  bool unite(int x, int y){
    x = find(x), y = find(y);
    if(x==y) return false;
    else if(rank[x]<rank[y]){
      par[x] = y;
      sizes[y] += sizes[x];
    } else if(rank[x]>rank[y]){
      par[y] = x;
      sizes[x] += sizes[y];
    } else{
      par[y] = x;
      rank[x] += 1;
      sizes[x] += sizes[y];
    }
    return true;
  }
  bool same(int x, int y) { return find(x) == find(y); }
  int size(int x) { return sizes[find(x)]; }
};

int main() {
  int n, m; cin>>n>>m;
  vector<pair<int, int>> ab(m);
  rep(i, m) {
    int a, b; cin>>a>>b; --a; --b;
    ab[i].first = a; ab[i].second = b;
  }

  vector<long long> ans(m);
  UnionFind uf(n);
  RREP(i, m, 0) {
    if (uf.find(ab[i].first) != uf.find(ab[i].second)) {
      ans[i] = (long long)uf.size(ab[i].first) * uf.size(ab[i].second);
      uf.unite(ab[i].first, ab[i].second);
    }
  }
  rep(i, m-1) ans[i+1] += ans[i];
  rep(i, m) print(ans[i]);

  return 0;
}